services/agents/channel_test.go

```diff
@@ -76,188 +80,277 @@ func setup(t *testing.T, connect func(agent.Agent_ConnectServer) error, expected
+		return io.EOF // any error without GRPCStatus() method
 	}
 
-	channel, _, teardown := setup(t, connect, status.Error(codes.Unimplemented, "Test error"))
+	stream, _, teardown := setup(t, connect, status.Error(codes.Canceled, context.Canceled.Error()))
 	defer teardown(t)
 
-	resp := channel.SendRequest(&agent.AgentMessage_QanData{
-		QanData: new(agent.QANDataRequest),
+	err := stream.Send(&agent.AgentMessage{
+		Id: 1,
+		Payload: &agent.AgentMessage_QanData{
+			QanData: new(agent.QANDataRequest),
+		},
 	})
-	assert.Nil(t, resp)
+	assert.NoError(t, err)
+
+	_, err = stream.Recv()
+	assert.Equal(t, status.Error(codes.Unknown, "EOF"), err)
 }
 
-func TestAgentClosesConnection(t *testing.T) {
-	connect := func(stream agent.Agent_ConnectServer) error {
-		err := stream.Send(&agent.ServerMessage{
-			Id: 1,
-			Payload: &agent.ServerMessage_Ping{
-				Ping: new(agent.PingRequest),
-			},
+func TestAgentClosesStream(t *testing.T) {
+	connect := func(ch *Channel) error { //nolint:unparam
+		msg := ch.SendRequest(&agent.ServerMessage_Ping{
+			Ping: new(agent.PingRequest),
 		})
-		assert.NoError(t, err)
+		assert.Nil(t, msg)
 
-		msg, err := stream.Recv()
-		assert.Equal(t, status.Error(codes.Canceled, "context canceled"), err)
+		assert.Nil(t, <-ch.Requests())
+		return nil
+	}
+
+	stream, _, teardown := setup(t, connect, io.EOF)
+	defer teardown(t)
+
+	msg, err := stream.Recv()
+	require.NoError(t, err)
+	assert.NotNil(t, msg)
+
+	err = stream.CloseSend()
+	assert.NoError(t, err)
+}
+
+func TestAgentClosesConnection(t *testing.T) {
+	connect := func(ch *Channel) error { //nolint:unparam
+		msg := ch.SendRequest(&agent.ServerMessage_Ping{
+			Ping: new(agent.PingRequest),
+		})
 		assert.Nil(t, msg)
 
+		assert.Nil(t, <-ch.Requests())
 		return nil
 	}
 
-	errClientConnClosing := status.Error(codes.Canceled, "grpc: the client connection is closing") // == grpc.ErrClientConnClosing
-	errConnClosing := status.Error(codes.Unavailable, "transport is closing")
-	channel, cc, teardown := setup(t, connect, errClientConnClosing, errConnClosing)
+	stream, cc, teardown := setup(t, connect, status.Error(codes.Canceled, context.Canceled.Error()))
 	defer teardown(t)
 
-	req := <-channel.Requests()
-	ping := req.GetPing()
-	assert.NotNil(t, ping)
+	msg, err := stream.Recv()
+	require.NoError(t, err)
+	assert.NotNil(t, msg)
 
-	assert.NoError(t, cc.Close())
+	err = cc.Close()
+	assert.NoError(t, err)
 }
 
-func TestUnexpectedMessageFromServer(t *testing.T) {
-	connect := func(stream agent.Agent_ConnectServer) error {
-		// this message triggers "no subscriber for ID" error
-		err := stream.Send(&agent.ServerMessage{
-			Id: 111,
-			Payload: &agent.ServerMessage_QanData{
-				QanData: new(agent.QANDataResponse),
-			},
+func TestUnexpectedResponseFromAgent(t *testing.T) {
+	connect := func(ch *Channel) error { //nolint:unparam
+		// after receiving unexpected response, channel is closed
+		msg := ch.SendRequest(&agent.ServerMessage_Ping{
```

Comment:
why don't we return error here?

Reply:
Two reasons. First, we can't really give a user a useful error there: was it send error? was it channel error, so message _maybe_ was sent, but not handled? or maybe it was handled, but we were not able to get a response? Second, we still need users to `Requests()` until it is closed, and not break earlier on error.